### ISSUE TITLE
Fails to build from 20240110 -> 20240118: ld: error: undefined symbol: pthread_create

### ISSUE THREAD

**Issue by nunotexbsd** [exult/exult#436]
FreeBSD clang version 17.0.6

```
/bin/sh ./libtool  --tag=CXX   --mode=link c++ -std=c++17  -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -I/usr/local/include   -fstack-protector-strong  -o exult exult-actions.o exult-actorio.o exult-actors.o exult-args.o exult-browser.o exult-cheat.o exult-cheat_screen.o exult-combat.o exult-dir.o exult-drag.o exult-effects.o exult-exult.o exult-exultmenu.o exult-game.o exult-gameclk.o exult-gamedat.o exult-gamemap.o exult-gamerend.o exult-gamewin.o exult-istring.o exult-keys.o exult-keyactions.o exult-menulist.o exult-monsters.o exult-mouse.o exult-npcnear.o exult-npctime.o exult-palette.o exult-party.o exult-paths.o exult-readnpcs.o exult-schedule.o exult-shapeid.o exult-touchui.o exult-tqueue.o exult-txtscroll.o exult-verify.o exult-version.o pathfinder/libpathfinder.la audio/libaudio.la audio/midi_drivers/libmididrv.la audio/midi_drivers/timidity/libtimidity.la conf/libconf.la imagewin/libimagewin.la server/libserver.la gamemgr/libgamemgr.la flic/libflic.la usecode/libusecode.la shapes/libshapes.la shapes/shapeinf/libshapeinf.la objs/libobjs.la gumps/libgumps.la files/libu7file.la files/sha1/libsha1.la files/zip/libminizip.la  -lpng -L/usr/local/lib -lSDL2  -L/usr/local/lib -lX11 -lXext  -lz   -L/usr/local/lib -logg -lvorbis -lvorbisfile    -L/usr/local/lib
libtool: link: c++ -std=c++17 -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -I/usr/local/include -fstack-protector-strong -o exult exult-actions.o exult-actorio.o exult-actors.o exult-args.o exult-browser.o exult-cheat.o exult-cheat_screen.o exult-combat.o exult-dir.o exult-drag.o exult-effects.o exult-exult.o exult-exultmenu.o exult-game.o exult-gameclk.o exult-gamedat.o exult-gamemap.o exult-gamerend.o exult-gamewin.o exult-istring.o exult-keys.o exult-keyactions.o exult-menulist.o exult-monsters.o exult-mouse.o exult-npcnear.o exult-npctime.o exult-palette.o exult-party.o exult-paths.o exult-readnpcs.o exult-schedule.o exult-shapeid.o exult-touchui.o exult-tqueue.o exult-txtscroll.o exult-verify.o exult-version.o  pathfinder/.libs/libpathfinder.a -L/usr/local/lib audio/.libs/libaudio.a audio/midi_drivers/.libs/libmididrv.a audio/midi_drivers/timidity/.libs/libtimidity.a conf/.libs/libconf.a imagewin/.libs/libimagewin.a server/.libs/libserver.a gamemgr/.libs/libgamemgr.a flic/.libs/libflic.a usecode/.libs/libusecode.a shapes/.libs/libshapes.a shapes/shapeinf/.libs/libshapeinf.a objs/.libs/libobjs.a gumps/.libs/libgumps.a files/.libs/libu7file.a files/sha1/.libs/libsha1.a files/zip/.libs/libminizip.a -lpng -lSDL2 -lX11 -lXext -lz -logg -lvorbis -lvorbisfile
ld: error: undefined symbol: pthread_create
>>> referenced by LowLevelMidiDriver.cpp
>>>               LowLevelMidiDriver.o:(std::__1::thread::thread<int (&)(LowLevelMidiDriver*), LowLevelMidiDriver*, void>(int (&)(LowLevelMidiDriver*), LowLevelMidiDriver*&&)) in archive audio/midi_drivers/.libs/libmididrv.a
c++: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[3]: *** [Makefile:1063: exult] Error 1
```

**Comment by wench** [comment]
That really doesn't make much sense. but apparently clang in freebsd doesn't automatically link pthreads when using C++11 threads. This is not a problem for any other platform, threading support from C++11 shouldn't need special build options

**Comment by nunotexbsd** [comment]
@wench

Build is fixed when passing `LDFLAGS+= -pthread` on port framework.
Thinking if there's a better way to fix this.

**Comment by Dragon-Baroque** [comment]
There is a case statement on `$host_os` near the beginning of `configure.ac`, with `freebsd` as a possible choice.

Could you test whether it would be enough to add there a `SYSLIBS="-lpthread"` like we do for Solaris or Darwin ?

**Comment by wench** [comment]
This issue has me thinking that while I'm messing with the multithreaded code that I should isolate it and only compile it in for platforms with hardware midi drivers that need it but that wouldn't help with this issue because the mu;ltithreaded code is needed by the UnixSeq midi driver that would get compiled in FreeBSD builds though there is an argument to be made that all midi drivers should be controllable by compile flags but that is getting really off topic for this issue.

**Comment by nunotexbsd** [comment]
@Dragon-Baroque 

It works perfectly!

```

--- configure.ac.orig   2024-01-29 19:58:59 UTC
+++ configure.ac
@@ -78,6 +78,7 @@ case "$host_os" in
                AC_DEFINE(FREEBSD, 1, [Using FreeBSD])
                AC_MSG_RESULT([X11 (FreeBSD)])
                CXXFLAGS="$CXXFLAGS -I/usr/local/include"
+               SYSLIBS="-lpthread"
                ;;
        netbsd* )
                WINDOWING_SYSTEM="-DXWIN"
```

**Comment by Dragon-Baroque** [comment]
Added commit 38e8b8d1 to master ( with the `"-lphtread"` variant like the other SYSLIBS in `configure.ac` ).

**Comment by nunotexbsd** [comment]
Thank you!

**Comment by Dragon-Baroque** [comment]
> It works with either `-pthread` or `-lpthread`.

@nunotexbsd, you raised a good point.
It seems that `-pthread` is an option to the compiler, whereas `-lpthread` goes to the linker only,
it seems that `-pthread` causes `-lpthread` but not vice versa, and that `-pthread` is the _preferred_ way.
Then, `-pthread` should go into `CXXFLAGS`, not into `SYSLIBS`, like this, from the commit 38e8b8d1 :
```
                WINDOWING_SYSTEM="-DXWIN"
                AC_DEFINE(FREEBSD, 1, [Using FreeBSD])
                AC_MSG_RESULT([X11 (FreeBSD)])
-               CXXFLAGS="$CXXFLAGS -I/usr/local/include"
-               SYSLIBS="-lpthread"
+               CXXFLAGS="$CXXFLAGS -I/usr/local/include -pthread"
                ;;
        netbsd* )
                WINDOWING_SYSTEM="-DXWIN"
```
Would you please try it ?

Side question, please : there is a choice for `netbsd` just below the choice for `freebsd`, are NetBSD and FreeBSD related ? Should the change made to FreeBSD be made to NetBSD too ?

@wench, I read your statement 

> That really doesn't make much sense. but apparently clang in freebsd doesn't automatically link pthreads when using C++11 threads.

meaning that you expect GCC / CLang to automatically cause `-pthread` ( not `-lphthread`, I suppose ) when using C++11 threads, but that CLang on FreeBSD does not, is that correct ?

**Comment by nunotexbsd** [comment]
> @nunotexbsd, you raised a good point. It seems that `-pthread` is an option to the compiler, whereas `-lpthread` goes to the linker only, it seems that `-pthread` causes `-lpthread` but not vice versa, and that `-pthread` is the _preferred_ way. Then, `-pthread` should go into `CXXFLAGS`, not into `SYSLIBS`, like this, from the commit [38e8b8d](https://github.com/exult/exult/commit/38e8b8d140102ea9ec3632f618af7e73a43955df) :
> 
> ```
> -               CXXFLAGS="$CXXFLAGS -I/usr/local/include"
> -               SYSLIBS="-lpthread"
> +               CXXFLAGS="$CXXFLAGS -I/usr/local/include -pthread"
> ```
> 
> Would you please try it ?

It builds ok and run test is ok too. If you think that -pthread is more correct than -lpthread, please commit it.

> 
> Side question, please : there is a choice for `netbsd` just below the choice for `freebsd`, are NetBSD and FreeBSD related ? Should the change made to FreeBSD be made to NetBSD too ?

NetBSD pkgsrc exult port is at 1.6 version. They do not follow exult-devel. My opinion is to not add pthread to netbsd target.

We share same origin on BSD 4.4 and trhough time some things have diverged.
Nice picture: https://en.wikipedia.org/wiki/Berkeley_Software_Distribution#/media/File:Unix_history-simple.svg

Thanks!

**Comment by marzojr** [comment]
Reopening for discussion.

I think that the proper fix would be to use `-pthread` on all POSIX systems using pthreads for c++ threads, even those that normally use glibc and don't "need" it.

**Comment by Dragon-Baroque** [comment]
> I think that the proper fix would be to use `-pthread` on all POSIX systems using pthreads for c++ threads, even those that normally use glibc and don't "need" it.

And systems that compile with GCC or CLang because those two support `-pthread`.

In the `configure.ac`, that should mean, in the `$host_os` case statement :

- Linux and FreeBSD, certainly
- Android and Darwin/MacOS, subject to check that the GitHub actions pass,
- Cygwin, MinGW, yes, although these are placeholders for the Windows build actually been done with `Makefile.mingw` instead,
- But neither NetBSD, OpenBSD or Solaris ?

**Comment by marzojr** [comment]
Linux, Android, Darwin/MacOS, NetBSD, OpenBSD, FreeBSD, and Solaris are all POSIX systems, and they all have POSIX-compliant pthreads implementations. So all of those should be using `-pthread`.

I looked at a a list of [c/c++ compilers](https://en.wikipedia.org/wiki/List_of_compilers#C++_compilers) that support c++17, and they fall into a few groups:
* GCC or clang
* Based on clang/llvm
* Can be invoked using GCC/clang flags
* Are mostly for Windows software, and use Windows threads instead of pthreads (MSVC)

So might as well just add `-pthread` on them all. `-pthread` Has additional multithread semantics, and makes GCC/clang do things like use a thread-local errno, making it thread-safe.

The [GCC doc page](https://gcc.gnu.org/onlinedocs/gcc-4.8.5/libstdc++/manual/manual/using_concurrency.html) is illuminating:
> The library strives to be thread-safe when all of the following conditions are met:
> ...
> **Requisite command-line flags are used for atomic operations and threading.** Examples of this include -pthread and -march=native, although specifics vary depending on the host environment. See [Machine Dependent Options](http://gcc.gnu.org/onlinedocs/gcc/Option-Summary.html).

Meaning that using `-pthread` is **required** for libstdc++ to be thread-safe when GCC's threading model is POSIX. I have now to doble-check if this is true for MinGW version as well...

**Comment by marzojr** [comment]
Additional information:
* GCC page for [Solaris](https://gcc.gnu.org/onlinedocs/gcc-5.5.0/gcc/Solaris-2-Options.html) options
* [This comment](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=16519#c3) for a documentation issue for GCC:
    > Proposed doc patch that would categorize -pthread as a generic linker option.  I think "GNU/Linux Options" is the wrong place because it's also supported by all the BSD variants, Darwin, AIX64, HP-UX, plus x86 MinGW and Cygwin targets.  If you think this is wrong, speak up.

**Comment by marzojr** [comment]
Hm, this might be interesting:
[VMs](https://github.com/vmactions) for use in Github Actions for various OSes that are not natively supported.

**Comment by Dragon-Baroque** [comment]
> [VMs](https://github.com/vmactions) for use in Github Actions for various OSes that are not natively supported.

Quite interesting. To create a new workflow, do we just create the ci-...yml file or is there something more ?

@nunotexbsd, I was precisely considering adding a workflow for FreeBSD, should I :

- Build normally like in Linux : Clone the Git tree, perform `autoreconf`, `configure` and `make`, or
- Install the Ports, including the exult-devel game port, refresh the Git tree, `make package`, or
- Use Poudriere ?

**Comment by nunotexbsd** [comment]
>     * Build normally like in Linux : Clone the Git tree, perform `autoreconf`, `configure` and `make`, or
I never been in upstream side but if my program have support for windows, linux and freebsd, then tests should be run and build directly from source code. Any adjustments should be done by upstream for each OS/case supported. This way tests automation will be possible for new releases.

(Hope that I'm making sense here)
 
>     * Install the Ports, including the exult-devel game port, refresh the Git tree, `make package`, or
Why not? This way you will work directly with our ports framework and no need to adjust configs/patching since port is already doing it's job of porting.
`man ports`will show usefull comands to build ports from source.

>     * Use Poudriere ?
Yes, this way we build, install, create packages in a clean enviroment with jails. That is what project uses for building official packages and for testing porposes.

See how it goes and if you need a more in depth conversation, you can e-mail me.

**Comment by marzojr** [comment]
> Quite interesting. To create a new workflow, do we just create the ci-...yml file or is there something more ?
 
The docs in the repositories seem to imply it would be simple like that. I want to find if there are any projects using them to see their yml files and see if there is anything special that needs to be done to get an error status.

I will play around with the VMs over the weekend to see how easy it would be to setup, but I will need some assistance with *BSD stuff as I never actually used any of them.

On the plus side:
> The default shell in FreeBSD is `csh`, if you want to use `sh` to execute the `run` script, please set `usesh` to `true`.

I guess we will be seeing how portable our shell scripts are 🙂

**Comment by Dragon-Baroque** [comment]
> but I will need some assistance with *BSD stuff as I never actually used any of them.

I have played with a FreeBSD VM, up to being able to build Exult. My experience is :

- Different options for `ps` and `ls` : `ps -efH` and `ls -AlF` in Linux -> `ps -aud` and `ls -lahF` in FreeBSD, 
- Different package installer, compared to `dnf install -y` in Fedora and `apt-fast install -y` in Ubuntu, FreeBSD uses `pkg install -y`.

The task is to get the list of dependent packages in their FreeBSD names. I have it, I shall post that very soon here...

Then get the same list for the other *BSD, and for Solaris. We may live with some missing Audio drivers - FreeBSD is already missing `MT32Emu` - but we definitely need SDL, and GTK+ 3 if we build Studio.

Now, how could you convince GitHub to run _only one_ workflow ?

**Comment by Dragon-Baroque** [comment]
FreeBSD and probably other *BSD store the packages into `/usr/local/` instead of `/usr/`.
The `configure.ac` macro `PKG_CHECK_MODULES` and its command line equivalent `pkgconf` track this properly, but there are some hardcoded libraries such as `-lpng` in `configure.ac` and the various `Makefile.am`.
The Ports build system in FreeBSD overcomes this by adding some options to `./configure` to ensure that `/usr/local/lib` is present  to all link steps. But a native ( non Ports ) build fails.

@DominusExult, I need your help. Could you please check that `pkgconf --libs libpng` answers properly on macOS ? The answer does not matter, only the fact that it answers something correct, not the `Package libpng was not found in the pkg-config search path.`.
And please the same with `pkgconf --libs zlib` `... x11` `... xext` `... fluidsynth` `... mt32emu` and `... alsa` - for those that are installed.

**Comment by wench** [comment]
> > but I will need some assistance with *BSD stuff as I never actually used any of them.
> 
> I have played with a FreeBSD VM, up to being able to build Exult. My experience is :
> 
> * Different options for `ps` and `ls` : `ps -efH` and `ls -AlF` in Linux -> `ps -aud` and `ls -lahF` in FreeBSD,
> * Different package installer, compared to `dnf install -y` in Fedora and `apt-fast install -y` in Ubuntu, FreeBSD uses `pkg install -y`.
> 
> The task is to get the list of dependent packages in their FreeBSD names. I have it, I shall post that very soon here...
> 
> Then get the same list for the other *BSD, and for Solaris. We may live with some missing Audio drivers - FreeBSD is already missing `MT32Emu` - but we definitely need SDL, and GTK+ 3 if we build Studio.
> 
> Now, how could you convince GitHub to run _only one_ workflow ?

No  MT32Emu? I'm guessing this just because there is no ports package for it as it should be possible to compile it from source and install to the system. Its probably not appropriate for a workflow for exult to check out the Munt sources and the build it to test our build process

**Comment by Dragon-Baroque** [comment]
> Its probably not appropriate for a workflow for exult to check out the Munt sources and then build it to test our build process.

Formally, this may not be so straightforward. For example the iOS build _has to_ download and build Ogg and Vorbis, and it adds MT32Emu to them. Whether the ( future ) FreeBSD workflow downloads and build MT32Emu before Exult could well depend on the workflow execution time, which should be larger that for ( say ) macOS, since the FreeBSD system is provided as a Virtual Machine to Ubuntu.

On the other hand, nothing prevents _me_ to test it on my FreeBSD VM. Good point.

**Comment by wench** [comment]
I'd guess that Android Snapshot builds will be far longer, it takes over 40 minutes because it builds Exult twice.

**Comment by DominusExult** [comment]
> @DominusExult, I need your help. Could you please check that pkgconf --libs libpng answers properly on macOS ? The answer does not matter, only the fact that it answers something correct, not the Package libpng was not found in the pkg-config search path..
> And please the same with pkgconf --libs zlib ... x11 ... xext ... fluidsynth ... mt32emu and ... alsa - for those that are installed.

as long as I use `pkg-config` and not `pkgconf` all are found

> No MT32Emu? I'm guessing this just because there is no ports package for it as it should be possible to compile it from source and install to the system. Its probably not appropriate for a workflow for exult to check out the Munt sources and then build it to test our build process

Yes, munt is notoriously not represented in package managers. I had to work on getting it added to mingw ports and macports, failed at brew (because of a stupid technicality, when someone from brew needed munt they added it). As long as we know from @Dragon-Baroque that a self build munt works fine with Exult, we can neglect its missing in a CI, IMO.

> I'd guess that Android Snapshot builds will take far longer, it takes over 40 minutes because it builds Exult twice. Of course the Windows snapshots also build twice but Exult seems to take an extremely long time to build for Android

Android is actually getting build *four* times (armeabi-v7a, arm64-v8a, x86, x86_64), see https://github.com/exult/exult/blob/master/android/app/src/main/cpp/dependencies/CMakeLists.txt#L5

**Comment by wench** [comment]
4 Architectures with both release and debug builds would explain why so slow as its actually building 8 times. Around 6 minutes per build sounds about right.
Not sure why we build x86_64 and i686 for windows snapshots, surely just 32 bit builds would be fine

**Comment by wench** [comment]
Yeah I can see that but why? Is there any particular reason why 64 bit builds are being made beyond because we can? There are no compatibility problems running 32bit programs on x86_64 Windows. If there are concerns over workflow execution time, this is just something that I'd look at.

**Comment by Dragon-Baroque** [comment]
> I will play around with the VMs over the weekend to see how easy it would be to setup, but I will need some assistance with *BSD stuff as I never actually used any of them.

@marzo, I have created a `ci-freebsd.yml` in my branch `exult-build-freebsd` and with the proper `paths:` clauses a (draft) Pull Request #440 against that branch triggered that workflow only and `CodeQL`.

I placed the required `pkg install` for FreeBSD in the `prepare` phase, and all the build commands in the `run` phase of the FreeBSD VM, and it looks like the Exult build succeeded ...

**Comment by Dragon-Baroque** [comment]
Closed by PR #440
